lib/plugin-loader.js

```diff
@@ -752,71 +756,275 @@ PluginLoader.prototype = {
+    if (host && port) {
+      let url = 'http://' + host + ':' + port + '/server/agent/environment';
+    
+      return new Promise((resolve, reject) => { /* Obtains and stores environment information from agent */
+        http.get(url, (res) => {
+        const { statusCode } = res; // TODO: Check status code for bad status
+        const contentType = res.headers['content-type'];
+
+        res.setEncoding('utf8');
+        let rawData = '';
+        res.on('data', (chunk) => { rawData += chunk; });
+        res.on('end', () => {
+          try {
+            const parsedData = JSON.parse(rawData);
+            if (parsedData.agentVersion) {
+              agentComp.version = parsedData.agentVersion;
+            }
+            if (parsedData.arch) {
+              agentComp.cpu = parsedData.arch;
+            }
+            if (parsedData.os == undefined && parsedData.hardwareIdentifier && parsedData.osRelease) {
+              agentComp.os = "zos";
+              agentComp.osRelease = parsedData.osRelease;
+              agentComp.hardwareIdentifier = parsedData.hardwareIdentifier;
+            } else {
+              agentComp.os = parsedData.os;
+            }
+            resolve();
+          } catch (e) {
+            bootstrapLogger.severe(e.message);
+            resolve(); // We don't want to reject here. Error gets caught down stream
+          }
+        });
+      }).on('error', (e) => {
+        bootstrapLogger.severe(e.message);
+        resolve();
+      });
+    }).then(() => { /* Obtains and stores the endpoints exposed by the agent */
+
+      'http://' + host + ':' + port + '/server/agent/services';
+
+      return new Promise((resolve, reject) => { 
+        http.get(url, (res) => {
+        const { statusCode } = res; // TODO: Check status code for bad status
+        const contentType = res.headers['content-type'];
+
+        res.setEncoding('utf8');
+        let rawData = '';
+        res.on('data', (chunk) => { rawData += chunk; });
+        res.on('end', () => {
+          try {
+            const parsedData = JSON.parse(rawData);
+            if (parsedData.services) {
+              agentComp.endpoints = [];
+              for (let i = 0; i < parsedData.services.length; i++) {
+                if (parsedData.services[i].urlMask) {
+                  agentComp.endpoints.push(parsedData.services[i].urlMask);
+                }
+              }
+            }
+            resolve();
+          } catch (e) {
+            bootstrapLogger.severe(e.message);
+            resolve(); // We don't want to reject here. Error gets caught down stream
+          }
+        });
+        }).on('error', (e) => {
+          bootstrapLogger.severe(e.message);
+          resolve(); // We don't want to reject here. Error gets caught down stream
+        });
+      }).then(() => {
+        let obj = {};
+        obj[AGENT_COMP_ID] = agentComp;
+        obj[APP_SERVER_COMP_ID] = appServerComp;
```

Comment:
this is fine but we need some comment in this function to state the purpose:

before checking if dependencies are met, we must learn about the components that exist.
doing this is not formalized currently, so we currently have a block per component to learn about their capabilities, version, environment, etc. perhaps in the future zowe components could have metadata files and/or expected URLs for querying.